phoenix-core-client/src/main/java/org/apache/phoenix/cache/ServerMetadataCacheImpl.java

```diff
@@ -0,0 +1,165 @@
+/**
+ * This manages the cache for all the objects(data table, views, indexes) on each region server.
+ * Currently, it only stores LAST_DDL_TIMESTAMP in the cache.
+ */
+public class ServerMetadataCacheImpl implements ServerMetadataCache {
+
+    protected Configuration conf;
+    // key is the combination of <tenantID, schema name, table name>, value is the lastDDLTimestamp
+    protected final Cache<ImmutableBytesPtr, Long> lastDDLTimestampMap;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ServerMetadataCacheImpl.class);
+    private static final String PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS =
+            "phoenix.coprocessor.regionserver.cache.ttl.ms";
+    // Keeping default cache expiry for 30 mins since we won't have stale entry
+    // for more than 30 mins.
+    private static final long DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS
+            = 30 * 60 * 1000L; // 30 mins
+    private static final String PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE
+            = "phoenix.coprocessor.regionserver.cache.size";
+    private static final long DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE = 10000L;
+    private static volatile ServerMetadataCacheImpl cacheInstance;
+    private MetricsMetadataCachingSource metricsSource;
+
+    /**
+     * Creates/gets an instance of ServerMetadataCache.
+     *
+     * @param conf configuration
+     * @return cache
+     */
+    public static ServerMetadataCacheImpl getInstance(Configuration conf) {
+        ServerMetadataCacheImpl result = cacheInstance;
+        if (result == null) {
+            synchronized (ServerMetadataCacheImpl.class) {
+                result = cacheInstance;
+                if (result == null) {
+                    cacheInstance = result = new ServerMetadataCacheImpl(conf);
+                }
+            }
+        }
+        return result;
+    }
+
+    public ServerMetadataCacheImpl(Configuration conf) {
+        this.conf = HBaseConfiguration.create(conf);
+        this.metricsSource = MetricsPhoenixCoprocessorSourceFactory
+                                .getInstance().getMetadataCachingSource();
+        long maxTTL = conf.getLong(PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS,
+                DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS);
+        long maxSize = conf.getLong(PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE,
+                DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE);
+        lastDDLTimestampMap = CacheBuilder.newBuilder()
+                .removalListener((RemovalListener<ImmutableBytesPtr, Long>) notification -> {
+                    String key = notification.getKey().toString();
+                    LOGGER.debug("Expiring " + key + " because of "
+                            + notification.getCause().name());
+                })
+                // maximum number of entries this cache can handle.
+                .maximumSize(maxSize)
+                .expireAfterAccess(maxTTL, TimeUnit.MILLISECONDS)
```

Comment:
We should consider adding some jitter here otherwise all the RS will expire roughly around the same time.

Reply:
@tkhurana Access time is reset by a read operation and since clients can pick regionservers at random, can we expect expiry of entries to also be a bit random?

Reply:
@palashc Actually I don't see how this ttl will help. It will be reset on every access so for popular entries the ttl expiry will never kick in. And for less popular entries your size limit will take care of evicting them. I would recommend just keeping the size limit.

Reply:
That makes sense!